File: kalen/money.py
from abc import ABC, abstractmethod
class Expression(ABC):

    @abstractmethod
    def reduce(self, to: str):
        pass

    @abstractmethod
    def __add__(self, addend):
        pass
    # 抽象クラスにもtimesメソッドを設定
    @abstractmethod
    def __mul__(self, multiplier: int):
        pass

class Pair:
    def __init__(self, fromcurrency: str, to: str):
        self.fromcurrency = fromcurrency
        self.to = to
    def __eq__(self, something):
        pair = something
        return self.fromcurrency == pair.fromcurrency and self.to == pair.to
    
    def __hash__(self):
        return hash(0)

# ---------------------

class Bank:
    def __init__(self):
        self.rates = {}
    
    def reduce(self, source: Exception, to: str):
        if type(source) is Money:
            return source

        sum = source
        return sum.reduce(self, to)
    def addRate(self, fromcurrency: str, to: str, rate: int):
        curr_rate = Pair(fromcurrency, to)
        self.rates[curr_rate] = rate

    def rate(self, fromcurrency: str, to: str):
        curr_rate = Pair(fromcurrency, to)
        if fromcurrency == to:
            return 1
        return self.rates.get(curr_rate)

# ---------------------

class Sum(Expression):
    def __init__(self, augend, addend):
        self.__augend = augend
        self.__addend = addend
    @property
    def augend(self):
        return self.__augend
    @property
    def addend(self):
        return self.__addend
    
    def reduce(self, bank, to):
        amount = self.augend.reduce(bank, to).amount + self.addend.reduce(bank, to).amount
        return Money(amount, to)
    
    def __add__(self, addend):
        return Sum(self, addend)
    # timesメソッドを追加
    def __mul__(self, multiplier: int):
        return Sum(self.augend * multiplier, self.addend * multiplier)
    
# ---------------------

class Money(Expression):
    def __init__(self, amount, currency):
        self.__amount = amount
        self.__currency = currency
    
    @property
    def currency(self):
        return self.__currency

    @property
    def amount(self):
        return self.__amount
    
    @staticmethod
    def dollar(amount: int):
        return Money(amount, "USD")

    @staticmethod
    def franc(amount: int):
        return Money(amount, "CHF")

    def __eq__(self, object) -> bool:
        return self.__dict__ == object.__dict__

    def __add__(self, addend):
        return Sum(self, addend)

    def __mul__(self, multiplier: int):
        return Money(self.amount * multiplier, self.currency)
        
    def reduce(self, bank, to: str):
        rate = bank.rate(self.currency, to)
        return Money(self.amount / rate, to)
    

File: shino/src/main/java/org/codepolaris/tdd/Money.java
package org.codepolaris.tdd;

import org.codepolaris.tdd.utils.Currency;

import java.util.Objects;

public class Money implements Expression {
  protected int amount;
  protected String currency;

  Money(int amount, String currency) {
    this.amount = amount;
    this.currency = currency;
  }

  public Money times(int multiplier) {
    return new Money(amount * multiplier, currency());
  }

  public Expression plus(Money addeed) {
    return new Sum(this, addeed);
  }

  @Override
  public Money reduce(String to) {
    return this;
  }

  public String currency() {
    return currency;
  }

  public static Money dollar(int amount) {
    return new Money(amount, Currency.USD);
  }

  public static Money franc(int amount) {
    return new Money(amount, Currency.CHF);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o instanceof Money money) {
      return amount == money.amount && currency().equals(money.currency());
    }
    return false;
  }

  @Override
  public int hashCode() {
    return Objects.hash(amount);
  }

  @Override
  public String toString() {
    return amount + " " + currency;
  }
}
